examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1673 @@
+
+
+# Based on ResidualBlock from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/shared.py#L20
+class ResidualBlock(torch.nn.Module):
+    def __init__(self, fn: Callable):
+        super().__init__()
+        self.fn = fn
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return (self.fn(x) + x) / np.sqrt(2)
+
+
+# Based on make_block from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L64
+class DiscHeadBlock(torch.nn.Module):
+    """
+    StyleGAN-T block: SpectralConv1d => BatchNormLocal => LeakyReLU
+    """
+
+    def __init__(
+        self,
+        channels: int,
+        kernel_size: int,
+        num_groups: int = 8,
+        leaky_relu_neg_slope: float = 0.2,
+    ):
+        super().__init__()
+        self.channels = channels
+
+        self.conv = SpectralConv1d(
+            channels,
+            channels,
+            kernel_size=kernel_size,
+            padding=kernel_size // 2,
+            padding_mode="circular",
+        )
+        self.batch_norm = torch.nn.GroupNorm(num_groups, channels)
+        self.act_fn = torch.nn.LeakyReLU(leaky_relu_neg_slope, inplace=True)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.conv(x)
+        x = self.batch_norm(x)
+        x = self.act_fn(x)
+        return x
+
+
+# Based on DiscHead in the official StyleGAN-T implementation
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L78
+# TODO: implement image conditioning (see Section 3.2 of paper)
+class DiscriminatorHead(torch.nn.Module):
+    """
+    Implements a StyleGAN-T-style discriminator head.
+    """
+
+    def __init__(
+        self,
+        channels: int,
+        cond_embedding_dim: int,
+        cond_map_dim: int = 64,
+    ):
+        super().__init__()
+        self.channels = channels
+        self.cond_embedding_dim = cond_embedding_dim
+        self.cond_map_dim = cond_map_dim
+
+        self.input_block = DiscHeadBlock(channels, kernel_size=1)
+        self.resblock = ResidualBlock(DiscHeadBlock(channels, kernel_size=9))
+
+        if self.cond_embedding_dim > 0:
+            self.conditioning_map = torch.nn.Linear(self.cond_embedding_dim, cond_map_dim)
+            self.cls = SpectralConv1d(channels, cond_map_dim, kernel_size=1, padding=0)
+        else:
+            self.cls = SpectralConv1d(channels, 1, kernel_size=1, padding=0)
+
+    def forward(self, x: torch.Tensor, c: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.input_block(x)
+        hidden_states = self.resblock(hidden_states)
+        out = self.cls(hidden_states)
+
+        if self.cond_embedding_dim > 0:
+            c = self.conditioning_map(c).squeeze(-1)
```

Comment:
Seems like `c` could be made default to `None`?

Reply:
It should not be `None`. It the `stylegan-t` and `ADD` paper `c` is always used, it could be `clip text embed` or `clip image embed` or both

Reply:
If that is the case, then the condition to enter this codepath seems irrelevant.

Reply:
Yes, that should be removed.

Reply:
Removed.